devel/changes

```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+# Extract changelog for a specified version (or __NEXT__ if no version
+# specified).
+set -euo pipefail
+
+cd "$(dirname "$0")/.."
+
+version="${1:-__NEXT__}"
+export version
+
+perl -ne 'print if /^# \Q$ENV{version}\E( |$)/ ... (/^# / and exit)' CHANGES.md | tail -n +2
```

Comment:
_(non-blocking)_

Could you break this down to a perl newbie? So far I understand:

- `-n`: loop through lines in CHANGES.md
- `-e`: execute a line of perl
- `^# `: match start of string + `# `.
- `\Q$ENV{version}\E`: match the version exactly (escape chars from regex)
- `( |$)`: match ` ` or EOL

Reply:
I break it down below because learning is fun… but also maybe I shouldn't have just gone with my semi-cheeky example from another PR review comment and instead written a bit more verbose version in a language the team knows. :upside_down_face: :sweat_smile: Perhaps [one of these variants](https://gist.github.com/tsibley/3db7f8df1ec61d407354b0e6c36ba170) (probably _c_), for example.

Break down of:

    perl -ne 'print if /^# \Q$ENV{version}\E( |$)/ ... (/^# / and exit)' CHANGES.md

This is a Perl "one-liner" which loops through lines (`-n`) of input evaluating (`-e`) a bit of code for each line.  (Very much like `sed` or `awk`.)  The current line of input is implicitly assigned to Perl's global "topic variable", `$_`, which can be thought of as "the current thing the code is talking about".  For many functions, the topic is the default argument if no explicit argument is provided.

Input is from the file `CHANGES.md` in this case, but the command would read from stdin if no filenames were passed (or multiple files sequentially if multiple filenames were passed).

The code conditions a call to the core `print` function (a keyword) with a postfix `if` statement.  The `print` function when given no arguments will use the topic variable (i.e. the current line of input here).  Postfix `if` statements are identical to prefix `if` statements and are only syntactic sugar.

The `if`'s condition uses the [triple-dot range operator](https://perldoc.perl.org/perlop#Range-Operators), `...`, which maintains a boolean state based on its left-hand side (LHS) and right-hand side (RHS).  It evalutes to false initially, turns true when the LHS evalutes to true, and stays true until the RHS evalutes to false, at which point it turns false again (until the LHS turns true again, if ever).  (The range operator has origins in `sed` and `awk`, as does the idea of a topic line, which Perl extends to the topic variable more generally.)

The LHS of the range operator tests the topic against a regular expression, which you already grok.

The RHS of the range operator tests a compound expression which evalutes to false unless the topic matches the regular expression, in which case it exits the program early.

It might be illustrative to see an equivalent program increasingly "spelled out":

```perl
while (<ARGV>) {
    print if /^# \Q$ENV{version}\E( |$)/ ... (/^# / and exit);
}
```

```perl
my $found = 0;

while (<ARGV>) {
    if ($found) {
        exit if /^# /;
        print;
    }
    elsif (/^# \Q$ENV{version}\E( |$)/) {
        $found = 1;
        print;
    }
}
```

Note that this still uses the topic variable.  It too could be "spelled out" with explicit lexical variables, but the topic is an idiomatic part of Perl in many contexts.

```perl
my $found = 0;

while (my $line = <ARGV>) {
    if ($found) {
        exit if $line =~ /^# /;
        print $line;
    }
    elsif ($line =~ /^# \Q$ENV{version}\E( |$)/) {
        $found = 1;
        print $line;
    }
}
```

Ok, I'll stop there, as this is likely to be more than you wanted to know about Perl.